File: Arrays & Strings/151. Reverse Words in a String.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    string reverseWords(string s)
    {
        std::stack<std::string> words;

        std::string word = "";
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == ' ')
            {
                if (word != "")
                {
                    words.push(word);
                    word = "";
                }
            }
            else
            {
                word += s[i];
            }
        }
        words.push(word);

        std::string res = "";
        while (!words.empty())
        {
            res += words.top();
            words.pop();
            res += ' ';
        }
        res.pop_back();

        return res;
    }
};

int main()
{
    string s = "  hello world";
    Solution sol;
    cout << sol.reverseWords(s) << endl;

    return 0;
}

File: Arrays & Strings/151. Reverse Words in a String.py

# using stack
def reverserWords(s: str) -> str:
    if (len(s) == 0):
        return s

    # iterate through the words of the str
    # add the curr word to the stack
    stack = []
    s_s = s.split()
    
    for i in s_s:
        stack.append(i)
    
    # pop the elements and add to the res
    res = ""
    while (stack):
        res += stack.pop()
        res += " "
        
    return res


if __name__ == "__main__":
    s = "Hello World"
    print(reverserWords(s))
    